pandas/_libs/src/ujson/python/objToJSON.c

```diff
@@ -396,96 +396,115 @@ static PyObject *get_item(PyObject *obj, Py_ssize_t i) {
-    return PandasDateTimeStructToJSON(&dts, tc, outValue, _outLen);
+    // Note that get_datetime_iso_8601_strlen just gives a generic size
+    // for ISO string conversion, not the actual size used
+    *len = strlen(result);
+    return result;
 }
 
-static void *PyDateTimeToJSON(JSOBJ _obj, JSONTypeContext *tc, void *outValue,
-                              size_t *_outLen) {
+static npy_datetime NpyDateTimeToEpoch(npy_datetime dt, NPY_DATETIMEUNIT base) {
+    scaleNanosecToUnit(&dt, base);
+    return dt;
+}
+
+static char *PyDateTimeToIso(JSOBJ obj, JSONTypeContext *tc, size_t *len) {
     npy_datetimestruct dts;
-    PyDateTime_Date *obj = (PyDateTime_Date *)_obj;
+    int ret;
 
-    PRINTMARK();
+    if (!PyDateTime_Check(obj)) {
+        // TODO: raise TypeError
+    }
 
-    if (!convert_pydatetime_to_datetimestruct(obj, &dts)) {
-        PRINTMARK();
-        return PandasDateTimeStructToJSON(&dts, tc, outValue, _outLen);
-    } else {
+    ret = convert_pydatetime_to_datetimestruct(obj, &dts);
+    if (ret != 0) {
         if (!PyErr_Occurred()) {
             PyErr_SetString(PyExc_ValueError,
-                            "Could not convert datetime value to string");
+                            "Could not convert PyDateTime to numpy datetime");
         }
         ((JSONObjectEncoder *)tc->encoder)->errorMsg = "";
         return NULL;
     }
+
+    NPY_DATETIMEUNIT base = ((PyObjectEncoder *)tc->encoder)->datetimeUnit;
+    *len = (size_t)get_datetime_iso_8601_strlen(0, base);
+    char *result = PyObject_Malloc(*len);
+    ret = make_iso_8601_datetime(&dts, result, *len, base);
+
+    if (ret != 0) {
+        PRINTMARK();
+        PyErr_SetString(PyExc_ValueError,
+                        "Could not convert datetime value to string");
+        ((JSONObjectEncoder *)tc->encoder)->errorMsg = "";
+        PyObject_Free(result);
+        return NULL;
+    }
+
+    // Note that get_datetime_iso_8601_strlen just gives a generic size
+    // for ISO string conversion, not the actual size used
+    *len = strlen(result);
+    return result;
 }
 
-static void *NpyDatetime64ToJSON(JSOBJ _obj, JSONTypeContext *tc,
-                                 void *outValue, size_t *_outLen) {
+static npy_datetime PyDateTimeToEpoch(PyObject *obj, NPY_DATETIMEUNIT base) {
     npy_datetimestruct dts;
-    PRINTMARK();
+    int ret;
+
+    if (!PyDateTime_Check(obj)) {
+        // TODO: raise TypeError
+    }
 
-    pandas_datetime_to_datetimestruct((npy_datetime)GET_TC(tc)->longValue,
-                                      NPY_FR_ns, &dts);
-    return PandasDateTimeStructToJSON(&dts, tc, outValue, _outLen);
+    ret = convert_pydatetime_to_datetimestruct(obj, &dts);
+    if (ret != 0) {
+        if (!PyErr_Occurred()) {
+            PyErr_SetString(PyExc_ValueError,
+                            "Could not convert PyDateTime to numpy datetime");
+        }
+        // TODO: is setting errMsg required?
```

Comment:
as long as the obj is a valid pydatetime and no tzinfo, i dont think this should ever be reached, is it?

Reply:
Probably not in current state but I think generally with C should always be checking return code of these functions and raising if something goes wrong.

Reply:
fair enough.  i think in the np_datetime code i went in the other direction in an attempt to make it less verbose